File: elixir/boutique-inventory/lib/boutique_inventory.ex
defmodule BoutiqueInventory do
  def sort_by_price(inventory) do
    inventory |> Enum.sort_by(&(&1[:price]))
  end

  def with_missing_price(inventory) do
    inventory |> Enum.filter(&(!&1[:price]))
  end

  def update_names(inventory, old_word, new_word) do
    inventory |> Enum.map(fn %{name: name} = item ->
      %{item | name: String.replace(name, old_word, new_word)}
    end)
  end

  def increase_quantity(%{quantity_by_size: qbs} = item, count) do
    qbs = Map.new(qbs, fn {size, cur_count} -> {size, cur_count + count} end)
    %{item | quantity_by_size: qbs}
  end

  def total_quantity(%{quantity_by_size: qbs}) do
    qbs |> Map.values() |> Enum.reduce(0, &+/2)
  end
end

File: ruby/boutique-inventory/boutique_inventory.rb
class BoutiqueInventory
  CHEAP_THRESHOLD = 30.0

  def initialize(items)
    @items = items
  end

  def item_names
    items.map { |item| item[:name] }.sort
  end

  def cheap
    items.filter { |item| item[:price] < CHEAP_THRESHOLD }
  end

  def out_of_stock
    items.filter { |item| item[:quantity_by_size].empty? }
  end

  def stock_for_item(name)
    items.find { |item| item[:name] == name }[:quantity_by_size]
  end

  def total_stock
    items.map { |item| item[:quantity_by_size].values }.flatten.sum
  end

  private
  attr_reader :items
end
